### PR TITLE
[PWA] feat/185 sign-in sign-out

### LINKED ISSUE #185: [PWA] Login / Logout
- [x] User login
- [x]  User login/user-menu/logout styling
- [x] Add menu item in header (automatically shown for logged in user)
- [x] Menu item styling
- [x] Menu item styling fixes, esp. for mobile
- [x] User logout confirmation page
- [x] Ensure cart is reset after logout

- [x] Remove layout shift generated by user menu behavior
- [x] Shift user menu to the left to make design consistent with the original website

- Showing user name/welcome text in `.user-menu-link`: #279 

- Out of scope: #202 

Screenshot 2023-04-12 at 09 35 51
Screenshot 2023-04-12 at 09 34 54

### PR DESCRIPTION
**Pull request by dicagno**
Please always provide the [GitHub issue(s)](../issues) your PR is for, as well as test URLs where your change can be observed (before and after):

Fix #185 

Test URLs:
- https://feat-185-signin-signup--maidenform--hlxsites.hlx.page/

### CODE REVIEW

File: src/talons/useLogout.js
Comment (dicagno) [inline src/talons/useLogout.js:1]: FIles like these should be leftovers from the old directory structure. Will remove as part of the cleanup

File: blocks/commerce-logout/commerce-logout.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import __pwa_bundle_helper__ from '/scripts/pwa-dist/pwa-bundle-helper.js';
```
Comment (herzog31) [inline blocks/commerce-logout/commerce-logout.js:1]: Please add the file to `.eslintignore`.

File: blocks/header/header.css [hunk 2]
```diff
@@ -1,503 +1,603 @@
 header .header.block {
```
Comment (herzog31) [inline blocks/header/header.css:1]: There are some stylelint issues in this file.

File: blocks/header/header.js [hunk 3]
```diff
@@ -36,6 +36,35 @@ function clearAllTabIndex() {
   });
 }
 
+function getLoggedInToken() {
+  return localStorage.getItem('M2_VENIA_BROWSER_PERSISTENCE__signin_token');
```
Comment (herzog31) [inline blocks/header/header.js:40]: Please use this function instead, which also respects the TTL: https://github.com/hlxsites/maidenform/blob/issue/184/scripts/commerce.js#L395-L415

File: blocks/header/header.js [hunk 4]
```diff
@@ -36,6 +36,35 @@ function clearAllTabIndex() {
...
+      .classList
+      .remove('user-menu-open');
+    return;
+  }
+
+  document.getElementById('userMenuLink').classList.add('user-menu-link-open');
+  document.getElementById('userMenu').classList.add('user-menu-open');
+};
+
+const toggleUserMenu = () => {
```
Comment (herzog31) [inline blocks/header/header.js:60]: You can do `element.classList.toggle('className')` which should simplify the logic here.

File: blocks/header/header.js [hunk 5]
```diff
@@ -171,6 +208,15 @@ export default async function decorate(block) {
     hamburger.classList.add('nav-hamburger');
     hamburger.innerHTML = '<span class="icon icon-mobile-menu"></span>';
 
+    // TODO: fetch basic user data
+    const loggedInToken = getLoggedInToken();
+
+    // TODO: get from document table?
```
Comment (herzog31) [inline blocks/header/header.js:214]: Can be hardcoded, that's fine.

File: blocks/header/header.js [hunk 6]
```diff
@@ -171,6 +208,15 @@ export default async function decorate(block) {
     hamburger.classList.add('nav-hamburger');
     hamburger.innerHTML = '<span class="icon icon-mobile-menu"></span>';
 
+    // TODO: fetch basic user data
```
Comment (herzog31) [inline blocks/header/header.js:211]: What about this? The first name of the user should be displayed.

File: src/components/Button/button.module.css [hunk 7]
```diff
@@ -0,0 +1,12 @@
+.button-mf {
```
Comment (herzog31) [inline src/components/Button/button.module.css:1]: Where is this used?

File: src/components/TextInput/textInput.module.css [hunk 8]
```diff
@@ -0,0 +1,46 @@
+.input {
```
Comment (herzog31) [inline src/components/TextInput/textInput.module.css:1]: Where is this used?

File: src/reducers/user.js [hunk 9]
```diff
@@ -0,0 +1,103 @@
+import { handleActions } from 'redux-actions';
```
Comment (herzog31) [inline src/reducers/user.js:1]: This file is currently exactly the same as https://github.com/magento/pwa-studio/blob/99925fae9c45be8f2f3bfd41b577d5754181acb9/packages/peregrine/lib/store/reducers/user.js.

File: src/talons/useSignin.js [hunk 10]
```diff
@@ -0,0 +1,195 @@
+import { useCallback, useRef, useState, useMemo } from 'react';
```
Comment (herzog31) [inline src/talons/useSignin.js:1]: This also looks like the exact file like in https://github.com/magento/pwa-studio/blob/99925fae9c45be8f2f3bfd41b577d5754181acb9/packages/peregrine/lib/talons/SignIn/useSignIn.js or did I miss any change?

File: blocks/header/header.js (hunk 6 above)
Comment (dicagno) [reply blocks/header/header.js]: Due to the complexity of integrating the `setToken` with the feature of setting user data in a way that is accessible by Franklin, IIRC we agreed only to have a static text and to investigate this more in detail in the next task.
This is even more important now that a bug in "blocking redirect" - before any operation (including this "localStorage" setting) can be executed - arose.

File: src/talons/useSignin.js (hunk 10 above)
Comment (dicagno) [reply src/talons/useSignin.js:1]: I missed deleting it. It was intended to tweak some functionality and get the set user details feature working

File: src/reducers/user.js (hunk 9 above)
Comment (dicagno) [reply src/reducers/user.js:1]: Same as below

File: blocks/header/header.css [hunk 11]
```diff
@@ -1,503 +1,603 @@
...
+    }
+
+    header nav .nav-tools .user-menu-link {
+        font-size: 1.2rem;
+        text-decoration: underline;
+        margin-right: 6px;
+        display: block;
+    }
+
+    header nav .nav-tools .user-menu-link::after {
```
[the remark this answers is not in the document]
Comment (dicagno) [reply blocks/header/header.css:328]: That's true. To reduce the LS I will use svg chevrons instead, they are more stable than Unicode text chevrons

[the remark this answers is not in the document]
Comment (dicagno) [reply blocks/header/header.css:326]: solved

File: src/components/TextInput/textInput.module.css (hunk 8 above)
Comment (dicagno) [reply src/components/TextInput/textInput.module.css]: Cleaning up

**Comment by mhaack** [comment]
Tested on localhost since the captcha did not work for me on the test branch. LGTM. @hannessolo please have a look as well.

**Review by mhaack — APPROVED** [review]

**Review by hannessolo — APPROVED** [review]
Looks good from a code perspective

**Pull request merged**